core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -134,6 +129,22 @@ public StartRecoveryRequest getRequest() {
      * performs the recovery from the local engine to the target
      */
     public RecoveryResponse recoverToTarget() throws IOException {
+        cancellableThreads.execute(() -> runUnderOperationPermit(() -> {
+            final IndexShardRoutingTable routingTable = shard.getReplicationGroup().getRoutingTable();
+            ShardRouting targetShardRouting = routingTable.getByAllocationId(request.targetAllocationId());
+            if (targetShardRouting == null) {
+                logger.debug("delaying recovery of {} as it is not listed as assigned to target node {}", request.shardId(),
+                    request.targetNode());
+                throw new DelayRecoveryException("source node does not have the shard listed in its state as allocated on the node");
+            }
+            if (targetShardRouting.initializing() == false) {
```

Comment:
true, this should never happen unless we reuse replica aids. Either way, the primary should probably not care to validate this, as the replica is in charge of saying when a recovery should occur, and the primary has only as role to validate that it is properly tracking the replica (the specific state the replica is in does not matter).